### PR TITLE
EthrDIDResolver

### PR DESCRIPTION
**Pull request by joshuabell**
Work:
- Refactor to favor structs over classes
- EthrDID build out which is a test foundation for EthrDIDResolver
- Code base structure and organization updated
- Many dependencies coded out
- Introduction of Promises
- General code improvements

Notes:
- Promises will be favored over closure based callbacks while waiting for coroutines in Swift 5. Both promises and callbacks will be implemented with callbacks methods just wrapping the promise based implementations.

- Will reintroduce AlamoFire for network requests

### CODE REVIEW

File: uPortSDK/Classes/EthrDID/EthrDIDResolver.swift [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                }
+                
+                fulfill( "\(parsedResponse.result!)")
+            }).catch({ error in
+                reject( error )
+            })
+        }
+    }
+    
+    func lastChangedSynchronous( identity: String ) throws -> String {
```
Comment (pelle) [inline uPortSDK/Classes/EthrDID/EthrDIDResolver.swift]: Dumb question. These functions marked Synchronous. Are they actually synchronous? Do they stall the app when run?

Comment (joshuabell) [reply uPortSDK/Classes/EthrDID/EthrDIDResolver.swift]: Correct! Synchronous here means they will stall the thread/queue they are called on until the network call completes. Synchronous methods are meant to be called on a background thread/queue.

File: uPortSDK/Classes/EthrDID/EthrDIDResolver.swift [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+                    let key = "DIDAttributeChanged-\(name)-\(event.value.value.hexEncodedString())"
+                    let ( section, algorithm, rawType, encoding ) = try self.destructuredAttributeChanged(eventName: name)
+                    let type = try self.parseType( algorithm: algorithm, rawType: rawType )
+                    switch section {
+                    case "pub":
+                        delegateCount += 1
+                        let publicKeyOwner = "\(normalizedDid)#delegate-\(delegateCount)"
+                        var pk = PublicKeyEntry(id: publicKeyOwner, type: type, owner: normalizedDid)
+                        switch encoding {
+                        case "", "null", "hex":
```
Comment (mirceanis) [inline uPortSDK/Classes/EthrDID/EthrDIDResolver.swift]: There's an error here:

"Expression pattern of type 'String' cannot match values of type 'String?'"
I'm not sure if this is coming from my older xcode (9.2) or is a genuine swift error

The error is fixed when switching on a non-optional variable.

Comment (mirceanis) [reply uPortSDK/Classes/EthrDID/EthrDIDResolver.swift]: I pushed a small commit fixing this

**Pull request merged**